File: src/main/java/wordcount/WordCounter.java
package wordcount;

import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class WordCounter {
    private final Set<String> stopWords;

    public WordCounter() {
        stopWords = Set.of();
    }

    public WordCounter(List<String> stopWords) {
        this.stopWords = new HashSet<>(stopWords);
    }

    public int countNumberOfWords(Collection<String> words) {
        if (words == null) {
            return 0;
        }
        int count = 0;
        for (String word : words) {
            if (! stopWords.contains(word)) {
                count++;
            }
        }
        return count;
    }
}

File: src/main/kotlin/util/WordCounter.kt
package util

import data.WordWrapper

class WordCounter {
    fun countWords(words: List<WordWrapper>) = filterWords(words).size

    fun countWords(words: List<WordWrapper>, stopWords: List<WordWrapper>) =
        filterWords(words = words, stopWords = stopWords).size

    fun countUniqueWords(words: List<WordWrapper>, stopWords: List<WordWrapper>): Int {
        val uniqueWords = mutableMapOf<String, Int>()
        words.map {
            var count = 0
            if (uniqueWords.contains(it.value)) {
                count = uniqueWords[it.value] ?: 0
                uniqueWords.replace(it.value, count.inc())
            } else {
                uniqueWords.put(it.value, count.inc())
            }
        }
        return words.count { word -> uniqueWords[word.value] == 1 && stopWords.none { word.value == it.value } && word.isAcceptedWord() }
    }

    fun averageWordLength(words: List<WordWrapper>): Float {
        var charCount = 0f
        words.map { word ->
            if (word.isAcceptedWord())
                charCount += word.value.length
        }
        return charCount.div(countWords(words).toFloat())
    }

    fun averageWordLength(words: List<WordWrapper>, stopWords: List<WordWrapper>): Float {
        var charCount = 0f
        words.map { word ->
            if (stopWords.none { word.value == it.value } && word.isAcceptedWord())
                charCount += word.value.length
        }
        return charCount.div(countWords(words, stopWords).toFloat())
    }


    fun filterWords(words: List<WordWrapper>) =
        words.filter { word -> word.isAcceptedWord() }

    fun filterWords(words: List<WordWrapper>, stopWords: List<WordWrapper>) =
        words.filter { word -> stopWords.none { word.value == it.value } && word.isAcceptedWord() }
}

fun String.splitToWords() =
    split(' ').map { WordWrapper(it) }